centml/compiler/server.py

```diff
@@ -1,20 +1,84 @@
+import os
+import pickle
+from http import HTTPStatus
+from typing import List
 import uvicorn
-from fastapi import FastAPI
-from pydantic import BaseModel
+from torch import Tensor
+from torch.fx import GraphModule
+from fastapi import FastAPI, UploadFile, HTTPException, BackgroundTasks
+from fastapi.responses import FileResponse
+from centml.compiler.server_compilation import hidet_backend_server, storage_path, CompilationStatus, dir_cleanup
+from centml.compiler import config_instance
+from fastapi.middleware.gzip import GZipMiddleware
+
 
 app = FastAPI()
+app.add_middleware(GZipMiddleware, minimum_size=1000)
 
-@app.post("/submit/{model_id}")
-async def submit_handler(model_id: int):
-	return {"message": "Compilation job submitted."}
 
 @app.get("/status/{model_id}")
-async def status_handler(model_id: int):
-	return {"message": "Compilation job finished."}
+async def status_handler(model_id: str):
+    if not os.path.isdir(os.path.join(storage_path, model_id)):
+        return {"status": CompilationStatus.NOT_FOUND}
+
+    if not os.path.isfile(os.path.join(storage_path, model_id, "cgraph.zip")):
+        return {"status": CompilationStatus.COMPILING}
+
+    if os.path.isfile(os.path.join(storage_path, model_id, "cgraph.zip")):
+        return {"status": CompilationStatus.DONE}
+
+    # Something is wrong if we get here
+    raise HTTPException(status_code=HTTPStatus.BAD_REQUEST, detail="Status check: invalid status state.")
+
+
+def background_compile(model_id: str, tfx_graph: GraphModule, example_inputs: List[Tensor]):
+    try:
+        # This will save the cgraph to {storage_path}/{model_id}/cgraph.zip
+        hidet_backend_server(tfx_graph, example_inputs, model_id)
+    except Exception:
+        dir_cleanup(model_id)
+
+
+@app.post("/submit/{model_id}")
+async def compile_model_handler(model_id: str, model: UploadFile, inputs: UploadFile, background_task: BackgroundTasks):
+    dir_cleanup(model_id)
+    os.makedirs(os.path.join(storage_path, model_id))
+
+    try:
+        tfx_contents = await model.read()
+        ei_contents = await inputs.read()
+    except Exception as e:
+        dir_cleanup(model_id)
+        raise HTTPException(
+            status_code=HTTPStatus.BAD_REQUEST, detail="Compilation: error reading serialized content."
+        ) from e
+    finally:
+        model.file.close()
```

Comment:
do we need to close `inputs.file` as well?

Reply:
Yes definitely. This was an oversight on my part. Thanks!